samples/matter/common/src/matter_init.h

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#pragma once
+
+#include <app/clusters/network-commissioning/network-commissioning.h>
+#include <lib/core/Optional.h>
+#include <lib/support/Variant.h>
+#include <platform/PlatformManager.h>
+#if CONFIG_CHIP_FACTORY_DATA
+#include <platform/nrfconnect/FactoryDataProvider.h>
+#else
+#include <platform/nrfconnect/DeviceInstanceInfoProviderImpl.h>
+#endif
+
+namespace Nordic
+{
+namespace Matter
+{
+	using ChipCustomInit = CHIP_ERROR (*)(void);
+	struct ChipInitData {
+		/** @brief Pointer to the user provided NetworkCommissioning instance. */
+		chip::app::Clusters::NetworkCommissioning::Instance *mNetworkingInstance{ nullptr };
+		/** @brief Optional custom initialization code that will be scheduled in the CHIP main event loop. */
+		ChipCustomInit mCustomInit{ nullptr };
+#if CONFIG_CHIP_FACTORY_DATA
```

Comment:
Yes, I have this in mind. I guess weather station is the good example of customized application and it gives an idea what might be desired from customer's perspective. I will try to align to that while integrating WS (which is an early stage for now.)

Reply:
Regarding the switchable variant, this was already considered and solved by explicitly feeding ChipInitData::mNetworkingInstance=nullptr in the Nordic::Matter::InitChipServer call.

Reply:
Yeah, but it's assigned only if not null:
```
			if (customInitData.mNetworkingInstance) {
				sCommissioningInstance = customInitData.mNetworkingInstance;
			}
```

Reply:
Ahh, right, thanks for the catch. The intention is to allow to explicitly register null networking instance (the sanity check is done later on anyway).

Reply:
The problem I see with accepting `CommissionableDataProvider` and other Matter interfaces instead of `FactoryDataProxy` is the fact that the latter contains the `Init()` method which is specific only to the nrfconnect FactoryData implementation (as a matter of fact most of other platforms do similar). Therefore I guess the initialization code must remain nrfconnect platform biased. As a workaround the user of MatterInit module can explicitly set the `mFactoryDataProvider` to nullptr and provide the customized implementation by directly using Matter API in the callback, e.g:

ReturnErrorOnFailure(Nordic::Matter::InitServer(
		MatterEventHandler, chip::MakeOptional<Nordic::Matter::InitData>(Nordic::Matter::InitData{
					    .mFactoryDataProvider = nullptr, .mPostServerInitClbk = [] {
						    chip::DeviceLayer::SetDeviceInstanceInfoProvider(cusomInfoProvider);
						    chip::Credentials::SetDeviceAttestationCredentialsProvider(customAttProvider);
                                                    chip::DeviceLayer::SetCommissionableDataProvider(customCommProvider);
						    return CHIP_NO_ERROR;
					    } })));

Reply:
I see, I guess then a user would need to call `Init` on their own. I don't have a good solution if we want to keep `Init` inside the common module...